module/zfs/bpobj.c

```diff
@@ -242,119 +277,171 @@ bpobj_iterate_impl(bpobj_t *bpo, bpobj_itor_t func, void *arg, dmu_tx_t *tx,
 		ASSERT3U(offset, >=, dbuf->db_offset);
 		ASSERT3U(offset, <, dbuf->db_offset + dbuf->db_size);
 
-		bparray = dbuf->db_data;
-		bp = &bparray[blkoff];
+		blkptr_t *bparray = dbuf->db_data;
+		blkptr_t *bp = &bparray[blkoff];
 		err = func(arg, bp, tx);
 		if (err)
 			break;
+
 		if (free) {
-			bpo->bpo_phys->bpo_bytes -=
-			    bp_get_dsize_sync(dmu_objset_spa(bpo->bpo_os), bp);
+			spa_t *spa = dmu_objset_spa(bpo->bpo_os);
+			freed += bp_get_dsize_sync(spa, bp);
 			ASSERT3S(bpo->bpo_phys->bpo_bytes, >=, 0);
 			if (bpo->bpo_havecomp) {
-				bpo->bpo_phys->bpo_comp -= BP_GET_PSIZE(bp);
-				bpo->bpo_phys->bpo_uncomp -= BP_GET_UCSIZE(bp);
+				comp_freed += BP_GET_PSIZE(bp);
+				uncomp_freed += BP_GET_UCSIZE(bp);
 			}
 			bpo->bpo_phys->bpo_num_blkptrs--;
 			ASSERT3S(bpo->bpo_phys->bpo_num_blkptrs, >=, 0);
 		}
 	}
+	if (free)
+		propagate_space_reduction(bpi, freed, comp_freed,
+		    uncomp_freed, tx);
 	if (dbuf) {
 		dmu_buf_rele(dbuf, FTAG);
 		dbuf = NULL;
 	}
 	if (free) {
-		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os, bpo->bpo_object,
-		    (i + 1) * sizeof (blkptr_t), DMU_OBJECT_END, tx));
+		VERIFY3U(0, ==, dmu_free_range(bpo->bpo_os,
+		    bpo->bpo_object, (i + 1) * sizeof (blkptr_t),
+		    DMU_OBJECT_END, tx));
 	}
-	if (err || !bpo->bpo_havesubobj || bpo->bpo_phys->bpo_subobjs == 0)
-		goto out;
+	return (err);
+}
 
-	ASSERT(bpo->bpo_havecomp);
-	err = dmu_object_info(bpo->bpo_os, bpo->bpo_phys->bpo_subobjs, &doi);
-	if (err) {
-		mutex_exit(&bpo->bpo_lock);
-		return (err);
+static void
+bpobj_iterate_cleanup(list_t *stack, int err)
+{
+	bpobj_t *bpo;
+
+	for (bpobj_info_t *bpi_cleanup = list_head(stack); bpi_cleanup != NULL;
+	    bpi_cleanup = list_next(stack, bpi_cleanup)) {
```

Comment:
It looks like this is using `bpi_cleanup` after it's been freed.  Is this code exercised by the test cases?  Or maybe kmem_free() doesn't poison the memory on ZoL (while testing)?

I think the usual way of doing this would be
```
while ((bpi = list_remove_head(stack)) != NULL) {
  ...
  kmem_free(bpi);
}